Source/santa-driver/SantaPrefixTree.cc

```diff
@@ -0,0 +1,238 @@
+
+      // If this is the end, mark the node as a prefix.
+      if (i + 1 == len) {
+        LOGD("Added prefix: %s", prefix);
+        new_node->isPrefix = true;
+      }
+
+      node->children[value] = (void *)new_node;
+      ++node_count_;
+      branched = true;
+
+      // Downgrade the exclusive lock
+      lck_rw_lock_exclusive_to_shared(spt_lock_);
+    } else if (i + 1 == len) {
+      // If the child does exist and it is the end...
+      // Set the new, higher prefix and prune the now dead nodes.
+
+      if (!lck_rw_lock_shared_to_exclusive(spt_lock_)) {
+        lck_rw_lock_exclusive(spt_lock_);
+      }
+
+      PruneNode((SantaPrefixNode *)node->children[value]);
+
+      SantaPrefixNode *new_node = new SantaPrefixNode();
+      new_node->isPrefix = true;
+
+      node->children[value] = (void *)new_node;
+      ++node_count_;
+
+      LOGD("Added prefix: %s", prefix);
+
+      lck_rw_lock_exclusive_to_shared(spt_lock_);
+    }
+
+    // Get ready for the next iteration.
+    node = (SantaPrefixNode *)node->children[value];
+  }
+
+  if (node_count) *node_count = node_count_;
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return kIOReturnSuccess;
+}
+
+bool SantaPrefixTree::HasPrefix(const char *string) {
+  lck_rw_lock_shared(spt_lock_);
+
+  auto found = false;
+
+  SantaPrefixNode *node = root_;
+
+  // TODO(bur): Avoid using strlen().
+  size_t len = strlen(string);
+  for (int i = 0; i < len; ++i) {
+    // Only process a byte at a time.
+    uint8_t value = string[i];
+
+    // Get the next char string.
+    node = (SantaPrefixNode *)node->children[value];
+
+    // If it doesn't exist in the tree, no match.
+    if (!node) break;
+
+    // If it does exist, is it a prefix?
+    if (node->isPrefix) {
+      found = true;
+      break;
+    }
+  }
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return found;
+}
+
+void SantaPrefixTree::Reset(uint32_t *node_count) {
+  lck_rw_lock_exclusive(spt_lock_);
+
+  PruneNode(root_);
+  root_ = new SantaPrefixNode();
+  ++node_count_;
+
+  if (node_count) *node_count = node_count_;
+  LOGD("Prefix tree reset");
+
+  lck_rw_unlock_exclusive(spt_lock_);
+}
+
+void SantaPrefixTree::PruneNode(SantaPrefixNode *target) {
+  if (!target) return;
+
+  // For deep trees a recursive approach will generate too many stack frames.
+  auto stack = OSArray::withCapacity(1);
+  auto seen = OSSet::withCapacity(1);
+
+  // Seed the "stack" with a starting node.
+  auto target_pointer = OSNumber::withNumber((uint64_t)(void *)target, 64);
+  stack->setObject(target_pointer);
+  target_pointer->release();
+
+  // Start at the target node and walk the tree to find all the sub-nodes.
+  while (stack->getCount()) {
+    auto pointer = OSDynamicCast(OSNumber, stack->getLastObject());
+    if (!pointer) {
+      LOGE("Unable complete prune!");
+      break;  // Bail walking the tree, but still delete any seen nodes.
+    }
+
+    seen->setObject(pointer);
+    stack->removeObject(stack->getCount() - 1);
```

Comment:
This might be premature - when you created this OSNumber object and added it to stack you released it, so the stack has the only reference to it. If you remove it from the stack the reference count should drop to 0 and be freed.

Reply:
I was adding it seen so it was retained. Got rid of seen, so it can now free the OSNumber as soon as it gets its value.